torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -0,0 +1,153 @@
+  // python object.
+  switch (messageType) {
+    case MessageType::SCRIPT_CALL: {
+      auto sc = static_cast<ScriptCall*>(rpc);
+
+      // sc is only alive within this block, use reference to avoid copy
+      auto& stack = sc->stackRef();
+      sc->op()->getOperation()(stack);
+
+      TORCH_INTERNAL_ASSERT(
+          stack.size() == 1,
+          "Return value of a builtin operator or a "
+          "TorchScript function should be a single IValue, got a vector of "
+          "size ",
+          stack.size());
+
+      return std::unique_ptr<ScriptResp>(
+          new ScriptResp(std::move(stack.front())));
+    }
+    case MessageType::PYTHON_CALL: {
+      auto pyCall = static_cast<PythonUDFCall*>(rpc);
+      auto payload = PythonRpcHandler::getInstance().generatePythonUDFResult(
+          pyCall->pickledPayload());
+      return std::unique_ptr<PythonUDFResp>(
+          new PythonUDFResp(std::move(payload)));
+    }
+    case MessageType::REMOTE_CALL: {
+      auto src = static_cast<ScriptRemoteCall*>(rpc);
+
+      auto rrefId = RRefId::fromIValue(src->retRRefId());
+      auto forkId = ForkId::fromIValue(src->retForkId());
+      TORCH_CHECK(rrefId != forkId, "Does not support remote call to self.");
+
+      auto& ctx = RRefContext::getInstance();
+      auto ownerRRef = ctx->getOrCreateOwnerRRef<IValue>(rrefId);
+
+      // TODO: make this asynchronous
+      // src is only alive within this block, use reference to avoid copy
+      auto& stack = src->stackRef();
+      src->op()->getOperation()(stack);
+      AT_ASSERT(
+          stack.size() == 1,
+          "Return value of a builtin operator or a "
+          "TorchScript function should be a single IValue, got a vector of "
+          "size ",
+          stack.size());
+
+      ownerRRef->setValue(std::move(stack.front()));
+      return nullptr;
+    }
+    case MessageType::RREF_FETCH_CALL: {
+      auto srf = static_cast<ScriptRRefFetchCall*>(rpc);
+      // TODO: make this asynchronous
+      std::shared_ptr<OwnerRRef<IValue>> rref =
+          RRefContext::getInstance()->getOrCreateOwnerRRef<IValue>(
+              RRefId::fromIValue(srf->value()));
+      return std::unique_ptr<ScriptRRefFetchRet>(
+          new ScriptRRefFetchRet(rref->getValue()));
+    }
+    case MessageType::RREF_USER_CREATE: {
+      auto sra = static_cast<ScriptRRefCreate*>(rpc);
+      RRefContext::getInstance()->addFork(sra->valueRef());
+      return nullptr;
+    }
+    case MessageType::RREF_USER_DELETE: {
+      auto srd = static_cast<ScriptRRefDelete*>(rpc);
+      RRefContext::getInstance()->delFork(srd->valueRef());
+      return nullptr;
+    }
+    case MessageType::MESSAGE_WITH_AUTOGRAD_REQ: {
+      auto rpcWithAutograd = static_cast<RpcWithAutograd*>(rpc);
+      const auto& autogradMetadata = rpcWithAutograd->autogradMetadata();
+
+      // Attach 'recv' autograd function.
+      DistAutogradContext* autogradContext = addRecvRpcBackward(
```

Comment:
I am a little worried of holding a raw pointer here. Any reason for not using a smart pointer?

Reply:
How are autograd containers going to be cleaned up? Are workers going to receive a message requesting to release a context_id? In what case, could it be that the "release autograd ctx" message arrives before this message finishes being processed?

Reply:
@aazzolini, when the autograd context is released on one node, we want to send an RPC to the other workers indicating that they should clean up their local context as well (see #25525 for details). The RPC would be sent in `DistAutogradContainer::releaseContext`. Will defer to @pritamdamania87 for the issue of message arrivals.

Reply:
if currentContext for this thread is set in addRecvRpcBackward(), we do not need to return this raw pointer

Reply:
> In what case, could it be that the "release autograd ctx" message arrives before this message finishes being processed?

In the normal case, this should not happen since the autograd context is cleaned up only after the backward pass is done. Although, in case of exceptions it is possible that the clean up message comes in while the autograd context is used. Doesn't seem like we can get away without shared_ptrs here.

I'm thinking of changing DistAutogradContainer.cpp to either always return a const reference to the context, or a shared_ptr if we need to mutate the context. Will do this in a separate PR.